### PR TITLE
[I256] Exponentiation. (5)

### PR DESCRIPTION
**Pull request by nlordell**
Implement exponentiation operations.

### Test Plan

New unit tests

### CODE REVIEW

File: src/int.rs [hunk 1]
```diff
@@ -471,6 +471,73 @@ impl I256 {
...
+    pub fn pow(self, exp: u32) -> Self {
+        handle_overflow(self.overflowing_pow(exp))
+    }
+
+    /// Raises self to the power of `exp`.
+    ///
+    /// Returns a tuple of the exponentiation along with a bool indicating
+    /// whether an overflow happened.
+    pub fn overflowing_pow(self, exp: u32) -> (Self, bool) {
+        let sign = match self.signum64().pow(exp & 1) {
```
Comment (fleupold) [inline src/int.rs]: 😭 🤕 🤯

I would have found logic like `if self.is_negative && exp % 2 ==> Sign Negative, else positive` much simpler to understand. But I guess this is correct too.

File: src/int.rs [hunk 2]
```diff
@@ -471,6 +471,73 @@ impl I256 {
...
+            Some(result)
+        }
+    }
+
+    /// Raises self to the power of `exp`, saturating at the numeric bounds
+    /// instead of overflowing.
+    pub fn saturating_pow(self, exp: u32) -> Self {
+        let (result, overflow) = self.overflowing_pow(exp);
+        if overflow {
+            return match self.signum64().pow(exp & 1) {
```
Comment (fleupold) [inline src/int.rs]: nit: maybe extract to common logic giving you the sign for an exp operation?

**Review by fleupold — APPROVED** [review]

File: src/int.rs (hunk 1 above)
Comment (nlordell) [reply src/int.rs]: Sorry! You’re right, that would be easier to understand.

**Review by e00E — APPROVED** [review]

**Comment by bh2smith** [comment]
This looks awesome! Why was it closed!? I love exponentiation!

**Comment by nlordell** [comment]
I merged all the 256-bit signed integer support in 1 PR, as I figured it didn’t make much sense to have partial support for it but it was too much code to review all at once.

Here it is: https://github.com/gnosis/ethcontract-rs/pull/231

**Comment by bh2smith** [comment]
Right, funny I was just looking at that one! So what does this mean for projects that already depend on this repo?

**Comment by nlordell** [comment]
It means they can now do signed integer math! 🥳

**Comment by bh2smith** [comment]
Right, but wasn't this already a possibility with primitive types available in parity-ethereum?

**Comment by nlordell** [comment]
>Right, but wasn't this already a possibility with primitive types available in parity-ethereum?

As far as I could tell, it only supported unsigned integers.

**Pull request closed without merging**